Add export capabilities to MSQ with SQL syntax (#15689)
apache/druid: docs/multi-stage-query/reference.md

### `EXTERN` Function

> Use the `EXTERN` function to read external data or write to an external source.

Reviewer: When we are writing to an external location, it is not a source. Something like the following might be better. 
```suggestion
Use the `EXTERN` function to read external data or write to an external location.
```

For more information, see [Read external data with EXTERN](concepts.md#read-external-data-with-extern).

#### `EXTERN` to export to a destination

> `EXTERN` can be used as a destination, which will export the data to the specified location and format. EXTERN when

Reviewer: ```suggestion
`EXTERN` can be used to specify a destination, where the data needs to be exported.  This variation of EXTERN requires one argument - (What the argument is)
```
minor nit: Also, from the docs, the format is specified via `AS` clause and not extern, therefore I have omitted that from the suggestion

used in this way accepts one argument. Please note that partitioning (`PARTITIONED BY`) and clustering (`CLUSTERED BY`)
is not currently supported with export statements.

INSERT statements and REPLACE statements are both supported with an `EXTERN` destination. The statements require an `AS`

> clause that determines the format.

Reviewer: ```suggestion
clause that specifies the format.
```

> Currently, only `CSV` is supported as a format.

Reviewer: ```suggestion
Only `CSV` format is supported at the moment
```

Export statements support the context parameter `rowsPerPage` for the number of rows in each exported file. The default value
is 100,000.

INSERT statements append the results to the existing files at the destination.
```sql
INSERT INTO
  EXTERN(<destination function>)
AS CSV
SELECT
  <column>
FROM <table>
```

REPLACE statements have an additional OVERWRITE clause. As partitioning is not yet supported, only `OVERWRITE ALL`

> is allowed. REPLACE deletes any existing files at the destination and creates new files with the results of the query.

Reviewer: "deletes any existing files" -> Does it clean up the directory, or just replace any files and paths conflicting with what the query generates the results for?

```sql
REPLACE INTO
  EXTERN(<destination function>)
AS CSV
OVERWRITE ALL
SELECT
  <column>
FROM <table>
```

> Exporting is currently supported to Amazon S3 storage. The S3 extension is required to be loaded for this.
> This can be done passing the function `S3()` as an argument to the `EXTERN` function.

Reviewer: ```suggestion
Exporting is currently supported for Amazon S3 storage. This can be done passing the function `S3()` as an argument to the `EXTERN` function. The `druid-s3-extensions` should be loaded.

```

> Exporting is currently supported for Amazon S3 storage. This can be done passing the function `S3()` as an argument to the `EXTERN` function. The `druid-s3-extensions` should be loaded.

Reviewer 2: I would like some more information on what this S3 function with the named is. Is it some special case or is it how we are settling on doing functions with named parameters. It it a SQL thing or a Calcite thing or a Druid thing?

I have at one point seen functions with names parameters be represented as

```
FN(x="a")
FN(x='a')
FN(x=>'a')
```

Where are all these variations coming from? Can there be quotes on the keys? Are they `"` or `'`? Can these functions also accept non-named (ordinal) parameters?

```sql
INSERT INTO

>   EXTERN(S3(bucket=<...>, prefix=<...>, tempDir=<...>))

Reviewer 2: I think this example would be clearer if you used syntax that would actually parse like:

`EXTERN(S3(bucket='s3://your_bucket, prefix='prefix/to/files', tempDir='/var'))`

Otherwise it is very hard to understand what actually needs to go in there. I have read these docs and I still do not understand if the values are quoted with `'` or `"`.

Author: Changed the example here

AS CSV
SELECT
  <column>
FROM <table>
```

##### S3

Export results to S3 by passing the function `S3()` as an argument to the `EXTERN` function. Note that this requires the `druid-s3-extensions`.
The `S3()` function is a Druid function that configures the connection. Arguments for `S3()` should be passed as named parameters with the value in single quotes like the following example:

```sql
INSERT INTO
  EXTERN(
    S3(bucket => 's3://your_bucket', prefix => 'prefix/to/files')
  )
AS CSV
SELECT
  <column>
FROM <table>
```

Supported arguments for the function:

| Parameter   | Required | Description                                                                                                                                                           | Default |
|-------------|----------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|
| `bucket`    | Yes      | The S3 bucket to which the files are exported to.                                                                                                                     | n/a     |

> | `prefix`    | Yes      | Path where the exported files would be created. The export query expects the destination to be empty. If the location includes other files, then the query will fail. | n/a     |

Reviewer 3: Why should mention that the bucket and prefix should match whatever the cluster admin has provided.

For more information, see [Read external data with EXTERN](concepts.md#read-external-data-with-extern).

#### `EXTERN` to export to a destination

`EXTERN` can be used to specify a destination, where the data needs to be exported.

> This variation of EXTERN requires one argument, the details of the destination as specified below.
> This variation additionally requires an `AS` clause to specify the format of the exported rows.

Reviewer 4: ```suggestion
This variation of EXTERN requires two arguments: the details of the destination and an `AS` clause to specify the format of the exported rows.
```

Author: The `AS` clause would not be an argument to extern, it's present elsewhere in the query

Reviewer 4: How about the change I just made?

> Only INSERT statements are supported with an `EXTERN` destination.
> Only `CSV` format is supported at the moment.
> Please note that partitioning (`PARTITIONED BY`) and clustering (`CLUSTERED BY`) is not currently supported with export statements.

Reviewer 4: ```suggestion
Keep the following in mind when using EXTERN to export rows:

- Only INSERT statements are supported.
- Only `CSV` format is supported as an export format.
- Partitioning (`PARTITIONED BY`) and clustering (`CLUSTERED BY`) aren't supported with export statements.
- You can export to Amazon S3 or local storage.
```

> Export statements support the context parameter `rowsPerPage` for the number of rows in each exported file. The default value

Reviewer 4: When you export data, use the `rowsPerPage` context parameter to control how many rows get exported. The default is 100,000.

> Export statements support the context parameter `rowsPerPage` for the number of rows in each exported file. The default value
> is 100,000.

Reviewer 4: ```suggestion
When you export data, use the `rowsPerPage` context parameter to control how many rows get exported. The default is 100,000.
```

> INSERT statements append the results to the existing files at the destination.
> ```sql

Reviewer 4: ```suggestion
INSERT statements append the results to the existing files at the destination.

```sql
```

Author: Removed this line. Added a new line `- The destination provided should be empty.` which is more accurate.

INSERT INTO
  EXTERN(<destination function>)
AS CSV
SELECT
  <column>
FROM <table>
```

> Exporting is currently supported for Amazon S3 storage and local storage.

Reviewer 4: ```suggestion
```

##### S3

> Exporting results to S3 can be done by passing the function `S3()` as an argument to the `EXTERN` function. The `druid-s3-extensions` should be loaded.

Reviewer 4: ```suggestion
Export results to S3 by passing the function `S3()` as an argument to the `EXTERN` function. Note that this requires the `druid-s3-extensions`.
```

> The `S3()` function is a druid function which configures the connection. Arguments to `S3()` should be passed as named parameters with the value in single quotes like the example below.

Reviewer 4: ```suggestion
The `S3()` function is a Druid function that configures the connection. Arguments for `S3()` should be passed as named parameters with the value in single quotes like the following example: 
```

```sql
INSERT INTO
  EXTERN(
    S3(bucket => 's3://your_bucket', prefix => 'prefix/to/files')
  )
AS CSV
SELECT
  <column>
FROM <table>
```

> Supported arguments to the function:

Reviewer 4: ```suggestion
Supported arguments for the function:
```

| Parameter   | Required | Description                                                                                                                                                                                                                        | Default |
|-------------|----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|
| `bucket`    | Yes      | The S3 bucket to which the files are exported to.                                                                                                                                                                                  | n/a     |

> | `prefix`    | Yes      | Path where the exported files would be created. The export query would expect the destination to be empty. If the location includes other files, then the query will fail.                                                         | n/a     |

Reviewer 4: ```suggestion
| `prefix`    | Yes      | Path where the exported files would be created. The export query expects the destination to be empty. If the location includes other files, the query will fail.                                                         | n/a     |
```

The following runtime parameters must be configured to export into an S3 destination:

| Runtime Parameter                            | Required | Description                                                                                                                                                                                                                        | Default |
|----------------------------------------------|----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----|
| `druid.export.storage.s3.tempSubDir`         | Yes      | Directory used to store temporary files required while uploading the data.                                                                                                                                                         | n/a |

> | `druid.export.storage.s3.allowedExportPaths` | Yes      | An array of S3 prefixes which are whitelisted as export destinations. Export query fail if the export destination does not match any of the configured prefixes. Example: `[\"s3://bucket1/export/\", \"s3://bucket2/export/\"]`     | n/a |

Reviewer 4: ```suggestion
| `druid.export.storage.s3.allowedExportPaths` | Yes      | An array of S3 prefixes that are whitelisted as export destinations. Export queries fail if the export destination does not match any of the configured prefixes. Example: `[\"s3://bucket1/export/\", \"s3://bucket2/export/\"]`     | n/a |
```

| `druid.export.storage.s3.maxRetry`           | No       | Defines the max number times to attempt S3 API calls to avoid failures due to transient errors.                                                                                                                                    | 10  |
| `druid.export.storage.s3.chunkSize`          | No       | Defines the size of each chunk to temporarily store in `tempDir`. The chunk size must be between 5 MiB and 5 GiB. A large chunk size reduces the API calls to S3, however it requires more disk space to store the temporary chunks. | 100MiB |

##### LOCAL

> Exporting is also supported to the local storage, which exports the results to the filesystem of the MSQ worker.

Reviewer 4: ```suggestion
You can export to the local storage, which exports the results to the filesystem of the MSQ worker.
```

> This is useful in a single node setup or for testing, and is not suitable for production use cases.

Reviewer 4: ```suggestion
This is useful in a single node setup or for testing but is not suitable for production use cases.
```

> This can be done by passing the function `LOCAL()` as an argument to the `EXTERN FUNCTION`.

Reviewer 4: ```suggestion
Export results to local storage by passing the function `LOCAL()` as an argument for the `EXTERN FUNCTION`. To use local storage as an export destination, the runtime property `druid.export.storage.baseDir` must be configured on the Indexer/Middle Manager.
The parameter provided to the `LOCAL()` function will be prefixed with this value when exporting to a local destination.
```

> Arguments to `LOCAL()` should be passed as named parameters with the value in single quotes like the example below.
> 
> To use local as an export destination, the runtime property `druid.export.storage.baseDir` must be configured on the indexer/middle manager.
> The parameter provided to the `LOCAL()` function will be prefixed with this value when exporting to a local destination.

Reviewer 4: ```suggestion

Arguments to `LOCAL()` should be passed as named parameters with the value in single quotes in the following example:

```

```sql
INSERT INTO
  EXTERN(
    local(exportPath => 'exportLocation/query1')
  )
AS CSV
SELECT
  <column>
FROM <table>
```

Supported arguments to the function:

| Parameter   | Required | Description                                                                                                                                                                                                                              | Default |
|-------------|--------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| --|

> | `exportPath`  | Yes | Subdirectory of `druid.export.storage.baseDir` used to as the destination to export the results to. The export query expects the destination to be empty. If the location includes other files or directories, then the query will fail. | n/a |

Reviewer 4: ```suggestion
| `exportPath`  | Yes | Subdirectory of `druid.export.storage.baseDir` used as the destination to export the results to. The export query expects the destination to be empty. If the location includes other files or directories, then the query will fail. | n/a |
```
